### PR TITLE
Make FermiWord compatible with jax

### PR DESCRIPTION
**Pull request by austingmhuang**
**Context:**
FermiWords and jax arrays weren't getting along. This fixes it.

**Description of the Change:**
Adds an additional conditional

**Benefits:**
Jax compatible

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-74796]

### CODE REVIEW

File: pennylane/fermi/fermionic.py [hunk 1]
```diff
@@ -147,15 +147,16 @@ def __add__(self, other):
...
 
         if isinstance(other, FermiSentence):
             return self_fs + other
 
         if isinstance(other, FermiWord):
             return self_fs + FermiSentence({other: 1.0})
 
-        if isinstance(other, (Number, ndarray)):
-            if isinstance(other, ndarray) and qml.math.size(other) > 1:
+        if isinstance(other, (Number, ndarray)) or use_jax:
```
Comment (soranjh) [inline pennylane/fermi/fermionic.py:158]: Cannot we just use `isinstance` to determine if the input data is a jax array?

File: pennylane/fermi/fermionic.py
Comment (soranjh) [inline pennylane/fermi/fermionic.py:1]: We also need tests for a jax objects. You can use `@pytest.mark.parametrize` and just add the jax array, Or just create a class for jax tests and add a few simple tests there.

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @austingmhuang, looks good overall. Just need a few simple tests and also if possible use isinstance as I commented.

File: pennylane/fermi/fermionic.py (hunk 1 above)
Comment (austingmhuang) [reply pennylane/fermi/fermionic.py:158]: No, you would have to import JAX to do that e.g.`import jax.array`

Better to use qml.math.get_interface in this case.

File: pennylane/fermi/fermionic.py [hunk 2]
```diff
@@ -147,15 +147,16 @@ def __add__(self, other):
         method"""
 
         self_fs = FermiSentence({self: 1.0})
+        use_jax = qml.math.get_interface(other) == "jax"
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:150]: You can perhaps combine "jax" and "numpy" here and then simplify the if condition below and at other places.

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Looks good! Make sure to include tests for these changes.

File: pennylane/fermi/fermionic.py [hunk 3]
```diff
@@ -147,15 +147,18 @@ def __add__(self, other):
...
         if isinstance(other, FermiSentence):
             return self_fs + other
 
         if isinstance(other, FermiWord):
             return self_fs + FermiSentence({other: 1.0})
 
-        if isinstance(other, (Number, ndarray)):
-            if isinstance(other, ndarray) and qml.math.size(other) > 1:
+        if is_jax_or_numpy:
+            if is_jax_or_numpy and qml.math.size(other) > 1:
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:161]: Why do we test `is_jax_or_numpy` two times?

File: pennylane/fermi/fermionic.py [hunk 4]
```diff
@@ -186,8 +192,8 @@ def __sub__(self, other):
             other_fs = FermiSentence(dict(zip(other.keys(), [-v for v in other.values()])))
             return self_fs + other_fs
 
-        if isinstance(other, (Number, ndarray)):
-            if isinstance(other, ndarray) and qml.math.size(other) > 1:
+        if is_jax_or_numpy:
+            if is_jax_or_numpy and qml.math.size(other) > 1:
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:196]: same as above.

Comment (austingmhuang) [reply pennylane/fermi/fermionic.py:196]: Oh, should be a different check inside

**Review by obliviateandsurrender — APPROVED** [review]
Looks good, remember to make all the tests pass!

File: pennylane/fermi/fermionic.py [hunk 5]
```diff
@@ -484,37 +470,31 @@ def __add__(self, other):
...
 
+        if not isinstance(other, (Number, FermiWord, FermiSentence)) and not hasattr(other, "size"):
+            raise TypeError(f"Cannot add {type(other)} to a FermiSentence.")
+        if qml.math.size(other) > 1:
+            raise ValueError(
+                f"Arithmetic Fermi operations can only accept an array of length 1, "
+                f"but received {other} of length {len(other)}"
+            )
+
         # ensure other is FermiSentence
```
Comment (soranjh) [inline pennylane/fermi/fermionic.py:481]: Please remove this comment, just noticed it now.

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @austingmhuang , looks great. My main comment is to use a TensorLike type check.

**Review by soranjh — APPROVED** [review]
Thanks @austingmhuang, looks pretty good.

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -86,6 +86,9 @@
   when called, will return a PLxPR representation of the input function.
   [(#6326)](https://github.com/PennyLaneAI/pennylane/pull/6326)
 
+* FermiWord and FermiSentence are now compatible with JAX arrays.
```
Comment (obliviateandsurrender) [inline doc/releases/changelog-dev.md:89]:
```suggestion
* `FermiWord` and `FermiSentence` are now compatible with JAX arrays.
```

File: pennylane/fermi/fermionic.py [hunk 7]
```diff
@@ -154,23 +152,19 @@ def __add__(self, other):
...
-        if isinstance(other, (Number, ndarray)):
-            if isinstance(other, ndarray) and qml.math.size(other) > 1:
-                raise ValueError(
-                    f"Arithmetic Fermi operations can only accept an array of length 1, "
-                    f"but received {other} of length {len(other)}"
-                )
-            return self_fs + FermiSentence({FermiWord({}): other})
+        if not isinstance(other, TensorLike):
+            raise TypeError(f"Cannot add {type(other)} to a FermiWord.")
+        if qml.math.size(other) > 1:
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:157]: Just add a newline between the `isisntance` check and the `size` assertion here and elsewhere.

**Review by obliviateandsurrender — APPROVED** [review]
Looks good, just a few final minor comments.

**Review by obliviateandsurrender — APPROVED** [review]
Looks good, just a few final minor comments.

**Pull request merged**